manuals/vmdkinfo.1
.Dd June 23, 2026
.Dt VMDKINFO 1
.Os
.Sh NAME
.Nm vmdkinfo
.Nd determines information about a VMware Virtual Disk (VMDK) file
.Sh SYNOPSIS
.Nm vmdkinfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm vmdkinfo
is a utility to determine information about a VMware Virtual Disk (VMDK) file
.Pp
.Nm vmdkinfo
is part of the
.Nm libvmdk
package.
.Nm libvmdk
is a library to access the VMware Virtual Disk (VMDK) format
.Pp
.Ar source
is the source image descriptor file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vmdkinfo file.vmdk
vmdkinfo 20130514

VMware Virtual Disk (VMDK) information:
	Disk type:			Monolithic sparse
	Media size:			8589934592 bytes
	Content identifier:		0x505dcf0a
	Parent content identifier:	0xffffffff

.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvmdk/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* fill block tree
* replace grain table with block tree
* replace extent table with block tree

* vmdkmount empty array on close not on free

* add support for the change tracking file?
* make sure descriptor file data is cleaned on close - also if no data files were set!

* make libvmdk_handle_open_extent_data_files_file_io_pool work
* make libvmdk_handle_open_extent_data_files use libvmdk_handle_open_extent_data_files_file_io_pool

* update manuals
  - update vmdkinfo.1
  - add vmdkmount.1
* handle ZERO extents?
* extent file add windows support (encoding)
  - add codepage functions + tests
  - libvmdk_extent_descriptor.c:629
  - libvmdk_descriptor_file.c:852
* implement extent file current offset ?
* implement libvmdk_extent_file_read_grain_group_element_data
* get grain data by grain table
  - remove unused functions from libvmdk_extent_table after refactor
* implement libvmdk_handle_set_maximum_number_of_open_handles
* add support for zero-grain table entry support

* code clean up:
  - support and locale functions
* add glob functions?
* debug output
  - descriptor file: disk database

API
* add functions to access storage media data
* add functions to access extents
* add functions to access grains

Tools:
* vmdkinfo:
  - add flag to have test to check if all the extent files are there?
  - same for all the parent extent files?
  - add option to print descriptor file data?
  - print dirty status
  - add byte size string support

Tests:
* open close
  - add parent support
* read
  - add parent support
* seek
  - add parent support
* add integrity tests

* Test images
  - vmdk flat
  - vmdk flat split (done)
  - Monolithic sparse (done)
  - 2GB extent sparse (done)
  - cowd sparse
  - cowd sparse split
  - compressed (done)
  - (compressed) with parent (done)

20090912
* see `git log' for more recent change log
* initial version based on liblnk
* worked on implementing vmdk and cowd sparse data files



vmdktools/vmdkinfo.c
/*
 * Shows information obtained from a VMware Virtual Disk (VMDK) file.
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "vmdktools_getopt.h"
#include "vmdktools_libcerror.h"
#include "vmdktools_libclocale.h"
#include "vmdktools_libcnotify.h"
#include "vmdktools_libvmdk.h"
#include "vmdktools_output.h"
#include "vmdktools_signal.h"
#include "vmdktools_unused.h"

info_handle_t *vmdkinfo_info_handle = NULL;
int vmdkinfo_abort                  = 0;

/* Signal handler for vmdkinfo
 */
void vmdkinfo_signal_handler(
      vmdktools_signal_t signal VMDKTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vmdkinfo_signal_handler";

	VMDKTOOLS_UNREFERENCED_PARAMETER( signal )

	vmdkinfo_abort = 1;

	if( vmdkinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     vmdkinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vmdkinfo to determine information about a VMware Virtual Disk (VMDK) file.";

	vmdktools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source descriptor file" },
	};
	system_character_t options_string[ 32 ];

	libvmdk_error_t *error                       = NULL;
	system_character_t * const *source_filenames = NULL;
	char *program                                = "vmdkinfo";
	system_integer_t option                      = 0;
	int number_of_options                        = (int) ( sizeof( options ) / sizeof( vmdktools_option_t ) );
	int number_of_source_filenames               = 0;
	int verbose                                  = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vmdktools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vmdktools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vmdktools_output_version_fprint(
	 stdout,
	 program );

	if( vmdktools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vmdktools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vmdktools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				vmdktools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vmdktools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source image.\n" );

		vmdktools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source_filenames           = &( argv[ optind ] );
	number_of_source_filenames = argc - optind;

	libcnotify_verbose_set(
	 verbose );
	libvmdk_notify_set_stream(
	 stderr,
	 NULL );
	libvmdk_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &vmdkinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open_input(
	     vmdkinfo_info_handle,
	     source_filenames,
	     number_of_source_filenames,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file(s).\n" );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     vmdkinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     vmdkinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &vmdkinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( vmdkinfo_info_handle != NULL )
	{
		info_handle_free(
		 &vmdkinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

